src/Debug/TraceableDomainEventDispatcher.php

```diff
@@ -0,0 +1,123 @@
+use Biig\Component\Domain\Event\DomainEventDispatcherInterface;
+use Biig\Component\Domain\Exception\InvalidArgumentException;
+use Biig\Component\Domain\Model\ModelInterface;
+use Biig\Component\Domain\Rule\DomainRuleInterface;
+use Biig\Component\Domain\Rule\PostPersistDomainRuleInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher;
+use Symfony\Component\EventDispatcher\Debug\WrappedListener;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class TraceableDomainEventDispatcher extends TraceableEventDispatcher implements DomainEventDispatcherInterface
+{
+    /**
+     * @var array
+     */
+    private $eventsFired;
+
+    /**
+     * @var DomainEventDispatcher
+     */
+    private $decorated;
+
+    /** @var array */
+    private $delayedListenersCalled;
+
+    /**
+     * TraceableDomainEventDispatcher constructor.
+     *
+     * @param DomainEventDispatcher $dispatcher
+     */
+    public function __construct(DomainEventDispatcher $dispatcher)
+    {
+        $this->eventsFired = [];
+        $this->delayedListenersCalled = [];
+        $this->decorated = $dispatcher;
+        parent::__construct($dispatcher, new Stopwatch(), new NullLogger());
+    }
+
+    public function addDomainRule(DomainRuleInterface $rule)
+    {
+        return $this->decorated->addDomainRule($rule);
+    }
+
+    public function addPostPersistDomainRuleInterface(PostPersistDomainRuleInterface $rule)
+    {
+        return $this->decorated->addPostPersistDomainRuleInterface($rule);
+    }
+
+    public function persistModel(ModelInterface $model)
+    {
+        /** @var DelayedListener $listener */
+        foreach ($this->decorated->getDelayedListeners() as $listener) {
+            $eventName = $listener->getEventName();
+
+            if ($listener->shouldOccur($model)) {
+                if (!$listener instanceof WrappedListener) {
+                    $listener = new WrappedDelayedListener($listener);
+                }
+
+                $this->delayedListenersCalled[] = $listener->getInfo($eventName);
+            }
+        }
+
+        return $this->decorated->persistModel($model);
+    }
+
+    /**
+     * Dispatches an event to all registered listeners.
+     *
+     * For BC with Symfony 4, the $eventName argument is not declared explicitly on the
+     * signature of the method. Implementations that are not bound by this BC constraint
+     * MUST declare it explicitly, as allowed by PHP.
+     *
+     * @param object      $event     The event to pass to the event handlers/listeners
+     * @param string|null $eventName The name of the event to dispatch. If not supplied,
+     *                               the class of $event should be used instead.
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return object The passed $event MUST be returned
+     */
+    public function dispatch($event, string $eventName = null)
+    {
+        if (!($event instanceof DomainEvent)) {
```

Comment:
Hum. AFAIK the domain dispatcher does not enforce events to be instances of `DomainEvent`. This could be a BC break (only for dev environment).

Reply:
You're right, I remove this